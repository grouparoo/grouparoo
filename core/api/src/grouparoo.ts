import { getCoreVersion, getNodeVersion } from "./utils/pluginDetails";

// have a shorter shutdown timeout
if (!process.env.ACTIONHERO_SHUTDOWN_TIMEOUT) {
  process.env.ACTIONHERO_SHUTDOWN_TIMEOUT = (15 * 1000).toString();
}

async function main() {
  const { Process, log } = await import("actionhero");

  log(
    `Starting Grouparoo ${getCoreVersion()} on node.js ${getNodeVersion()}`,
    "notice"
  );

  const app = new Process();

  app.registerProcessSignals((exitCode) => {
    log(
      `All Grouparoo tasks complete - exiting with code ${exitCode}`,
      "notice"
    );
    process.nextTick(() => process.exit(exitCode));
  });

  await app.start();
}

main();
