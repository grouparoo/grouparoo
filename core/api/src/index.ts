export { plugin } from "./modules/plugin";

export * from "./classes/plugin";
export * from "./classes/fileTransport";

export { App } from "./models/App";
export { ApiKey } from "./models/ApiKey";
export { Destination } from "./models/Destination";
export { DestinationGroup } from "./models/DestinationGroup";
export { DestinationGroupMembership } from "./models/DestinationGroupMembership";
export { Event } from "./models/Event";
export { EventData } from "./models/EventData";
export { Export } from "./models/Export";
export { File } from "./models/File";
export { Group } from "./models/Group";
export { GroupMember } from "./models/GroupMember";
export { GroupRule } from "./models/GroupRule";
export { Import } from "./models/Import";
export { Log } from "./models/Log";
export { Mapping } from "./models/Mapping";
export { Option } from "./models/Option";
export { Permission } from "./models/Permission";
export { Profile } from "./models/Profile";
export { ProfileProperty } from "./models/ProfileProperty";
export { ProfilePropertyRule } from "./models/ProfilePropertyRule";
export { ProfilePropertyRuleFilter } from "./models/ProfilePropertyRuleFilter";
export { Run } from "./models/Run";
export { Schedule } from "./models/Schedule";
export { Setting } from "./models/Setting";
export { Source } from "./models/Source";
export { Team } from "./models/Team";
export { TeamMember } from "./models/TeamMember";

export { waitForLock } from "./modules/locks";
