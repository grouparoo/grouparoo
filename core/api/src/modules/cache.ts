import { api, cache } from "actionhero";
import { waitForLock } from "./locks";

export interface ObjectCacheMethod {
  (
    argument: {
      objectGuid: string;
      cacheKey: string;
      cacheDurationMs?: number;
      lock?: boolean; // send false to no lock
      read?: boolean; // send false to not use cache
      write?: boolean; // send false to not write to cache
      passthru?: boolean; // send true to not read or write or lock
    },
    methodToGetValue: { (): Promise<any> }
  ): Promise<any>;
}

export interface ObjectCacheInvalidateMethod {
  (argument: { objectGuid: string }): Promise<void>;
}

async function getCacheValue({
  valueKey,
  read,
}): Promise<{ cached: boolean; value: any }> {
  if (!read) {
    return { cached: false, value: undefined };
  }
  try {
    const cached = await cache.load(valueKey);
    return { cached: true, value: cached.value };
  } catch (error) {
    const message = (error?.message || "").toString();
    if (message === "Object not found" || message === "Object expired") {
      return { cached: false, value: undefined };
    }
    throw error;
  }
}

function makeObjectKey({ objectGuid }) {
  objectGuid = (objectGuid || "").toString().trim();
  if (!objectGuid) {
    throw new Error(`objectGuid required`);
  }

  // TODO: validate if real guid?

  return `objectcache:${objectGuid}`;
}

function makeBaseKey({ objectGuid, cacheKey }): string {
  const objectKey = makeObjectKey({ objectGuid });

  cacheKey = (cacheKey || "").toString().trim();
  if (!cacheKey) {
    throw new Error(`cacheKey required`);
  }

  return `${objectKey}:${cacheKey}`;
}

function useRedis() {
  const running = !!api?.redis?.clients;
  if (!running && process.env.NODE_ENV === "test") {
    return false;
  }
  return true;
}

export const objectCacheInvalidate: ObjectCacheInvalidateMethod = async ({
  objectGuid,
}) => {
  if (!useRedis()) {
    return;
  }

  const client = cache.client();
  const objectKey = makeObjectKey({ objectGuid });
  const prefix = `${cache.redisPrefix}${objectKey}`;
  const keys = await client.keys(prefix + "*");
  const jobs = [];
  keys.forEach((key: string) => {
    jobs.push(client.del(key));
  });
  await Promise.all(jobs);
};

export const objectCache: ObjectCacheMethod = async (
  {
    objectGuid,
    cacheKey,
    cacheDurationMs = 1000 * 10,
    lock = true,
    read = true,
    write = true,
    passthru = false,
  },
  methodToGetValue
) => {
  if (!useRedis()) {
    passthru = true;
  }

  if (passthru) {
    read = false;
    write = false;
    lock = false;
  }

  const baseKey = makeBaseKey({ objectGuid, cacheKey });
  const valueKey = `${baseKey}:value`;
  const lockKey = `${baseKey}:lock`;

  let { value, cached } = await getCacheValue({ valueKey, read });
  if (cached) {
    return value;
  }

  let releaseLock: Function = null;
  try {
    if (lock) {
      releaseLock = (await waitForLock(lockKey)).releaseLock;

      let { value, cached } = await getCacheValue({ valueKey, read });
      if (cached) {
        return value;
      }
    }

    const value = await methodToGetValue();
    if (write) {
      await cache.save(valueKey, value, cacheDurationMs);
    }
    return value;
  } finally {
    if (releaseLock) {
      await releaseLock();
    }
  }
};
