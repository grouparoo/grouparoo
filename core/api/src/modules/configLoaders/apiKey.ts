import {
  ConfigurationObject,
  validateAndFormatGuid,
  codeConfigLockKey,
  logModel,
  validateConfigObjectKeys,
} from "../../classes/codeConfig";
import { ApiKey, Permission } from "../..";
import { Op } from "sequelize";

export async function loadApiKey(configObject: ConfigurationObject) {
  let isNew = false;

  const guid = await validateAndFormatGuid(ApiKey, configObject.id);
  validateConfigObjectKeys(ApiKey, configObject);

  let apiKey = await ApiKey.scope(null).findOne({
    where: { locked: codeConfigLockKey, guid },
  });
  if (!apiKey) {
    isNew = true;
    apiKey = await ApiKey.create({
      guid,
      locked: codeConfigLockKey,
      name: configObject.name,
    });
  }

  await apiKey.update({ name: configObject.name });

  if (
    configObject.options?.permissionAllRead !== undefined &&
    configObject.options?.permissionAllRead !== null
  ) {
    await apiKey.update({
      permissionAllRead: configObject.options.permissionAllRead,
    });
  }

  if (
    configObject.options?.permissionAllWrite !== undefined &&
    configObject.options?.permissionAllWrite !== null
  ) {
    await apiKey.update({
      permissionAllWrite: configObject.options.permissionAllWrite,
    });
  }

  if (configObject.permissions) {
    await apiKey.setPermissions(configObject.permissions);
  }

  await Permission.update(
    { locked: codeConfigLockKey },
    { where: { ownerGuid: apiKey.guid } }
  );

  logModel(apiKey, isNew ? "created" : "updated");
  return apiKey;
}

export async function deleteApiKeys(guids: string[]) {
  const apiKeys = await ApiKey.scope(null).findAll({
    where: { locked: codeConfigLockKey, guid: { [Op.notIn]: guids } },
  });

  for (const i in apiKeys) {
    await apiKeys[i].destroy();
    logModel(apiKeys[i], "deleted");
  }
}
