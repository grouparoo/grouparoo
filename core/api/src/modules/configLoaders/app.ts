import {
  ConfigurationObject,
  validateAndFormatGuid,
  extractNonNullParts,
  codeConfigLockKey,
  logModel,
} from "../../classes/codeConfig";
import { App } from "../..";
import { Op } from "sequelize";

export async function loadApp(configObject: ConfigurationObject) {
  let isNew = false;
  const guid = await validateAndFormatGuid(App, configObject.id);

  let app = await App.scope(null).findOne({
    where: { guid, locked: codeConfigLockKey },
  });
  if (!app) {
    isNew = true;
    app = await App.create({
      guid,
      locked: codeConfigLockKey,
      name: configObject.name,
      type: configObject.type,
    });
  }

  await app.update({ name: configObject.name });
  await app.setOptions(extractNonNullParts(configObject, "options"));

  const response = await app.test();
  if (!response.success) {
    throw new Error(
      `error testing app ${app.name} (${app.guid}) - ${response.error}`
    );
  }

  await app.update({ state: "ready" });
  logModel(app, isNew);
  return app;
}

export async function deleteApps(guids: string[]) {
  const apps = await App.scope(null).findAll({
    where: { locked: codeConfigLockKey, guid: { [Op.notIn]: guids } },
  });

  for (const i in apps) await apps[i].destroy();
}
