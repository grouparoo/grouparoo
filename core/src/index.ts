export { plugin } from "./modules/plugin";

export * from "./classes/plugin";
export * from "./classes/fileTransport";
export * from "./classes/configTemplate";
export { AuthenticatedAction } from "./classes/actions/authenticatedAction";
export { OptionallyAuthenticatedAction } from "./classes/actions/optionallyAuthenticatedAction";

export { App, SimpleAppOptions } from "./models/App";
export { ApiKey } from "./models/ApiKey";
export {
  Destination,
  SimpleDestinationOptions,
  DestinationSyncMode,
  DestinationSyncOperations,
} from "./models/Destination";
export { DestinationGroupMembership } from "./models/DestinationGroupMembership";
export { Export } from "./models/Export";
export { ExportProcessor } from "./models/ExportProcessor";
export { File } from "./models/File";
export { Group } from "./models/Group";
export { GroupMember } from "./models/GroupMember";
export { GroupRule } from "./models/GroupRule";
export { Import } from "./models/Import";
export { Log } from "./models/Log";
export { Notification } from "./models/Notification";
export { Mapping } from "./models/Mapping";
export { Option } from "./models/Option";
export { Permission } from "./models/Permission";
export { GrouparooRecord } from "./models/GrouparooRecord";
export { RecordProperty } from "./models/RecordProperty";
export {
  Property,
  PropertyTypes,
  PropertyFiltersWithKey,
  SimplePropertyOptions,
} from "./models/Property";
export { Filter } from "./models/Filter";
export { RecordMultipleAssociationShim } from "./models/RecordMultipleAssociationShim";
export { Run, HighWaterMark } from "./models/Run";
export {
  Schedule,
  SimpleScheduleOptions,
  PluginConnectionScheduleOption,
  ScheduleFiltersWithKey,
} from "./models/Schedule";
export { Session } from "./models/Session";
export { Setting } from "./models/Setting";
export { SetupStep } from "./models/SetupStep";
export { Source, SimpleSourceOptions, SourceMapping } from "./models/Source";
export { Team } from "./models/Team";
export { TeamMember } from "./models/TeamMember";

export { Status } from "./modules/status";
export { StatusMetric } from "./modules/statusReporters";
export { GrouparooCLI } from "./modules/cli";

export { waitForLock } from "./modules/locks";
export { Errors } from "./modules/errors";
export * from "./modules/cache";
export * from "./modules/tableSpeculation";
