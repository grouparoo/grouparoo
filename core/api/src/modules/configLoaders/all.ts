import { log, config, env } from "actionhero";
import path from "path";
import fs from "fs";
import glob from "glob";
import {
  ConfigurationObject,
  sortConfigurationObject,
} from "../../classes/codeConfig";
import { loadApp, deleteApps } from "./app";
import { loadSource, deleteSources } from "./source";
import {
  loadProfilePropertyRule,
  deleteProfilePropertyRules,
} from "./profilePropertyRule";
import { loadApiKey, deleteApiKeys } from "./apiKey";
import { loadTeam, deleteTeams } from "./team";
import { loadTeamMember, deleteTeamMembers } from "./teamMember";
import { loadGroup, deleteGroups } from "./group";
import { loadSchedule, deleteSchedules } from "./schedule";
import { loadSetting } from "./setting";
import { loadDestination, deleteDestinations } from "./destination";
import JSON5 from "json5";

interface SeenGuids {
  app: string[];
  source: string[];
  profilepropertyrule: string[];
  group: string[];
  schedule: string[];
  destination: string[];
  apikey: string[];
  team: string[];
  teammember: string[];
}

export async function loadConfigDirectory(configDir: string) {
  const globSearch = path.join(configDir, "**", "+(*.json|*.js)");
  const configFiles = glob.sync(globSearch);
  let configObjects: ConfigurationObject[] = [];
  for (const i in configFiles) {
    configObjects = configObjects.concat(await loadConfigFile(configFiles[i]));
  }

  if (configFiles.length > 0) {
    const seenGuids = await processConfigObjects(configObjects);
    await deleteLockedObjects(seenGuids);
  }
}

async function loadConfigFile(file: string): Promise<ConfigurationObject> {
  let payload;
  if (file.match(/\.json$/)) {
    payload = JSON5.parse(fs.readFileSync(file));
  } else {
    payload = require(file);
  }

  if (typeof payload === "function") payload = await payload(config);
  return payload;
}

async function processConfigObjects(configObjects: Array<ConfigurationObject>) {
  const seenGuids: SeenGuids = {
    app: [],
    source: [],
    profilepropertyrule: [],
    group: [],
    schedule: [],
    destination: [],
    apikey: [],
    team: [],
    teammember: [],
  };

  const sortedConfigObjects = configObjects.sort(sortConfigurationObject);

  for (const i in sortedConfigObjects) {
    const configObject = sortedConfigObjects[i];
    let klass = configObject.class.toLocaleLowerCase();
    let object;
    try {
      switch (klass) {
        case "setting":
          object = await loadSetting(configObject);
          break;
        case "app":
          object = await loadApp(configObject);
          break;
        case "source":
          object = await loadSource(configObject);
          break;
        case "profilepropertyrule":
          object = await loadProfilePropertyRule(configObject);
          break;
        case "group":
          object = await loadGroup(configObject);
          break;
        case "schedule":
          object = await loadSchedule(configObject);
          break;
        case "destination":
          object = await loadDestination(configObject);
          break;
        case "apikey":
          object = await loadApiKey(configObject);
          break;
        case "team":
          object = await loadTeam(configObject);
          break;
        case "teammember":
          object = await loadTeamMember(configObject);
          break;
        default:
          throw new Error(`unknown config object class: ${configObject.class}`);
      }
    } catch (error) {
      log(
        `error with config object: ${JSON.stringify(configObject)} - ${error}`,
        env === "test" ? "info" : "emerg"
      );
      throw error.original ? error.original : error;
    }

    if (klass !== "setting") seenGuids[klass].push(object.guid);
  }

  return seenGuids;
}

async function deleteLockedObjects(seenGuids: SeenGuids) {
  await deleteTeamMembers(seenGuids.teammember);
  await deleteTeams(seenGuids.team);
  await deleteApiKeys(seenGuids.apikey);
  await deleteDestinations(seenGuids.destination);
  await deleteSchedules(seenGuids.schedule);
  await deleteGroups(seenGuids.group);
  await deleteProfilePropertyRules(seenGuids.profilepropertyrule);
  await deleteSources(seenGuids.source);
  await deleteApps(seenGuids.app);
}
