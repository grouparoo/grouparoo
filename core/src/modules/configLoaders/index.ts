import { log, config, env } from "actionhero";
import path from "path";
import fs from "fs";
import glob from "glob";
import {
  ConfigurationObject,
  sortConfigurationObjects,
  validateConfigObjects,
  IdsByClass,
} from "../../classes/codeConfig";
import { loadApp, deleteApps } from "./app";
import { loadSource, deleteSources } from "./source";
import { loadProperty, deleteProperties } from "./property";
import { loadApiKey, deleteApiKeys } from "./apiKey";
import { loadTeam, deleteTeams } from "./team";
import { loadTeamMember, deleteTeamMembers } from "./teamMember";
import { loadGroup, deleteGroups } from "./group";
import { loadSchedule, deleteSchedules } from "./schedule";
import { loadSetting } from "./setting";
import { expandSyncTable } from "./syncTable";
import { loadDestination, deleteDestinations } from "./destination";
import JSON5 from "json5";
import { getParentPath } from "../../utils/pluginDetails";

export function getConfigDir() {
  const configDir =
    process.env.GROUPAROO_CONFIG_DIR || path.join(getParentPath(), "config");
  return configDir;
}

export async function loadConfigDirectory(
  configDir: string,
  externallyValidate: boolean = true
): Promise<{
  seenIds: IdsByClass;
  errors: string[];
  deletedIds: IdsByClass;
}> {
  let seenIds: IdsByClass = {};
  let deletedIds: IdsByClass = {};
  let errors: string[] = [];

  const configObjects = await loadConfigObjects(configDir);

  const response = await processConfigObjects(
    configObjects,
    externallyValidate
  );
  seenIds = response.seenIds;
  errors = response.errors;

  if (errors.length === 0) {
    deletedIds = await deleteLockedObjects(seenIds);
  }

  return { seenIds, errors, deletedIds };
}

export async function loadConfigObjects(
  configDir: string
): Promise<ConfigurationObject[]> {
  const globSearch = path.join(configDir, "**", "+(*.json|*.js)");
  const configFiles = glob.sync(globSearch);
  let configObjects: ConfigurationObject[] = [];
  for (const i in configFiles) {
    configObjects = configObjects.concat(await loadConfigFile(configFiles[i]));
  }
  configObjects = configObjects.filter((o) => Object.keys(o).length > 0); // skip empty files
  return configObjects;
}

async function loadConfigFile(file: string): Promise<ConfigurationObject> {
  let payload;
  if (file.match(/\.json$/)) {
    payload = JSON5.parse(fs.readFileSync(file).toString());
  } else {
    payload = require(file);
  }

  const payloadKeys = Object.keys(payload);
  if (payloadKeys.length === 1 && payloadKeys[0] === "default") {
    payload = payload.default;
  }

  if (typeof payload === "function") payload = await payload(config);
  return payload;
}

export async function processConfigObjects(
  configObjects: Array<ConfigurationObject>,
  externallyValidate: boolean,
  validate = false
): Promise<{ seenIds: IdsByClass; errors: string[] }> {
  const seenIds: IdsByClass = {
    app: [],
    source: [],
    property: [],
    group: [],
    schedule: [],
    destination: [],
    apikey: [],
    team: [],
    teammember: [],
  };
  const errors: string[] = [];

  configObjects = sortConfigurationObjects(configObjects);

  const { errors: validationErrors } = validateConfigObjects(configObjects);
  validationErrors.map((err) =>
    log(`[ config ] ${err}`, env === "test" ? "info" : "error")
  );
  errors.push(...validationErrors);

  if (errors.length > 0) return { seenIds, errors };

  for (const i in configObjects) {
    const configObject = configObjects[i];
    if (Object.keys(configObject).length === 0) continue;
    let klass = configObject?.class?.toLowerCase();
    let ids: IdsByClass;
    try {
      switch (klass) {
        case "setting":
          ids = await loadSetting(configObject, externallyValidate, validate);
          break;
        case "app":
          ids = await loadApp(configObject, externallyValidate, validate);
          break;
        case "source":
          ids = await loadSource(configObject, externallyValidate, validate);
          break;
        case "property":
          ids = await loadProperty(configObject, externallyValidate, validate);
          break;
        case "group":
          ids = await loadGroup(configObject, externallyValidate, validate);
          break;
        case "schedule":
          ids = await loadSchedule(configObject, externallyValidate, validate);
          break;
        case "destination":
          ids = await loadDestination(
            configObject,
            externallyValidate,
            validate
          );
          break;
        case "apikey":
          ids = await loadApiKey(configObject, externallyValidate, validate);
          break;
        case "team":
          ids = await loadTeam(configObject, externallyValidate, validate);
          break;
        case "teammember":
          ids = await loadTeamMember(
            configObject,
            externallyValidate,
            validate
          );
          break;
        case "synctable":
          const many = await expandSyncTable(
            configObject,
            externallyValidate,
            validate
          );
          const expanded = await processConfigObjects(
            many,
            externallyValidate,
            validate
          );
          ids = expanded.seenIds;
          errors.push(...expanded.errors);
          break;
        default:
          throw new Error(`unknown config object class: ${configObject.class}`);
      }
    } catch (error) {
      const errorMessage = `[ config ] error with ${configObject?.class} \`${
        configObject.key || configObject.name
      }\` (${configObject.id}): ${error}`;
      errors.push(errorMessage);
      log(errorMessage, "warning");
      continue;
    }

    // should set ids in all cases
    for (const className in ids) {
      const newIds = ids[className];
      seenIds[className].push(...newIds);
    }
  }

  return { seenIds, errors };
}

export async function deleteLockedObjects(seenIds) {
  const deletedIds: IdsByClass = {
    app: [],
    source: [],
    property: [],
    group: [],
    schedule: [],
    destination: [],
    apikey: [],
    team: [],
    teammember: [],
  };

  deletedIds["teammember"] = await deleteTeamMembers(seenIds.teammember);
  deletedIds["team"] = await deleteTeams(seenIds.team);
  deletedIds["apikey"] = await deleteApiKeys(seenIds.apikey);
  deletedIds["destination"] = await deleteDestinations(seenIds.destination);
  deletedIds["schedule"] = await deleteSchedules(seenIds.schedule);
  deletedIds["group"] = await deleteGroups(seenIds.group);
  deletedIds["property"] = await deleteProperties(seenIds.property);
  // might return a bootstrapped property, needs special processing
  const deletedSourceIds = await deleteSources(seenIds.source);
  deletedIds["source"] = deletedSourceIds.source;
  deletedSourceIds.property.map((id) => {
    if (!deletedIds["property"].includes(id)) deletedIds["property"].push(id);
  });
  // back to normal
  deletedIds["app"] = await deleteApps(seenIds.app);

  return deletedIds;
}
