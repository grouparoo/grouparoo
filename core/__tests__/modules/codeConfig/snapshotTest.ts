import { helper, relaxedSnapshot } from "@grouparoo/spec-helper";
import { api } from "actionhero";
import path from "path";
import { GrouparooPlugin } from "../../../src";
import { loadConfigDirectory } from "../../../src/modules/configLoaders";

describe("modules/codeConfig", () => {
  helper.grouparooTestServer({ enableTestPlugin: true, truncate: true });

  describe("with code config", () => {
    beforeAll(async () => {
      // manually run the initializer again after the server has started.
      // the test test-app plugin has been loaded
      api.codeConfig.allowLockedModelChanges = true;
      await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "initial")
      );

      // mock the test-plugin to return static results
      const testPlugin: GrouparooPlugin = api.plugins.plugins.find(
        (a) => a.name === "@grouparoo/test-plugin"
      );
      const testPluginConnection = testPlugin.connections.find(
        (c) => c.name === "test-plugin-import"
      );
      testPluginConnection.methods.recordProperty = async ({ property }) => {
        const data = {
          userId: [100],
          isVIP: [true],
          email: [`example@example.com`],
          firstName: ["Mario"],
          lastName: ["Mario"],
          ltv: [100.0],
          lastLoginAt: [new Date(0)],
          purchases: ["hat", "mushroom"],
          purchaseAmounts: [100, 200],
        };

        return data[property.key];
      };
    });

    test("getRecord will throw an error when not using a bootstrapped property", async () => {
      await expect(
        helper.getRecord({
          email: "example@example.com",
        })
      ).rejects.toThrow(
        /The arguments provided must contain key\/value pairs for one of the following: userId/
      );
    });

    test("a record snapshot can be tested with a bootstrapped property", async () => {
      const { record, snapshot } = await helper.getRecord({
        userId: 100,
      });

      // You can do snapshot testing
      expect(snapshot).toMatchSnapshot(relaxedSnapshot(snapshot));

      // Or you can test the properties of the snapshot directly
      expect(snapshot.properties.email.values).toEqual(["example@example.com"]);
      expect(snapshot.properties.userId.values).toEqual([100]);
      expect(snapshot.groups.length).toBe(1);
      expect(snapshot.groups[0].name).toBe("People with Email Addresses");
      expect(record.state).toBe("ready");
    });

    test("a record snapshot can be tested with multiple unique properties", async () => {
      const { record, snapshot } = await helper.getRecord({
        email: "example@example.com",
        userId: 100,
      });

      // You can do snapshot testing
      expect(snapshot).toMatchSnapshot(relaxedSnapshot(snapshot));

      // Or you can test the properties of the snapshot directly
      expect(snapshot.properties.email.values).toEqual(["example@example.com"]);
      expect(snapshot.properties.userId.values).toEqual([100]);
      expect(snapshot.groups.length).toBe(1);
      expect(snapshot.groups[0].name).toBe("People with Email Addresses");
      expect(record.state).toBe("ready");
    });
  });
});
