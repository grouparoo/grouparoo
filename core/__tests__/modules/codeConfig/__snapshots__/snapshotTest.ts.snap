// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/codeConfig with code config a profile snapshot can be tested 1`] = `
Object {
  "exports": Array [
    Object {
      "completedAt": null,
      "createdAt": null,
      "destination": null,
      "destinationName": "Test Destination",
      "errorLevel": undefined,
      "errorMessage": undefined,
      "force": true,
      "guid": null,
      "hasChanges": true,
      "mostRecent": false,
      "newGroups": Array [
        "Literally Everyone",
      ],
      "newProfileProperties": Object {
        "primary-id": 100,
        "secondary-id": "example@example.com",
      },
      "oldGroups": Array [],
      "oldProfileProperties": Object {},
      "profileGuid": Any<String>,
      "startedAt": Any<Number>,
      "toDelete": false,
    },
  ],
  "groups": Array [
    Object {
      "calculatedAt": null,
      "createdAt": Any<Number>,
      "guid": Any<String>,
      "locked": "config:code",
      "matchType": "all",
      "name": "People with Email Addresses",
      "nextCalculatedAt": Any<Number>,
      "profilesCount": 1,
      "rules": Array [
        Object {
          "key": "userId",
          "match": "null",
          "operation": Object {
            "description": "is not equal to",
            "op": "ne",
          },
          "relativeMatchDirection": null,
          "relativeMatchNumber": null,
          "relativeMatchUnit": null,
          "topLevel": false,
          "type": "integer",
        },
        Object {
          "key": "email",
          "match": "%@%",
          "operation": Object {
            "description": "is like (case sensitive)",
            "op": "like",
          },
          "relativeMatchDirection": null,
          "relativeMatchNumber": null,
          "relativeMatchUnit": null,
          "topLevel": false,
          "type": "email",
        },
      ],
      "state": "updating",
      "type": "calculated",
      "updatedAt": Any<Number>,
    },
  ],
  "properties": Object {
    "email": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "guid": Any<String>,
      "identifying": false,
      "isArray": false,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "email",
      "unique": true,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        "example@example.com",
      ],
    },
    "first name": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "guid": Any<String>,
      "identifying": false,
      "isArray": false,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        Object {},
      ],
    },
    "last name": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "guid": Any<String>,
      "identifying": false,
      "isArray": false,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        Object {},
      ],
    },
    "userId": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "guid": Any<String>,
      "identifying": true,
      "isArray": false,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "integer",
      "unique": true,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        100,
      ],
    },
  },
}
`;
